File: generate_slideshow_audio.py
#!/usr/bin/env python3
"""
Generate MP3 audio files for engineering slideshow using OpenAI TTS API.

Usage:
    export OPENAI_API_KEY="your-api-key-here"
    python generate_slideshow_audio.py
"""

import os
import json
import re
import time
from pathlib import Path
import urllib.request
import urllib.error

def extract_slideshow_scripts():
    """Extract all scripts from the slideshow HTML file."""
    with open('engineering_slideshow.html', 'r') as f:
        html_content = f.read()

    # Extract slideshowData JavaScript object
    match = re.search(r'const slideshowData = ({[\s\S]*?});', html_content)
    if not match:
        raise ValueError("Could not find slideshowData in HTML")

    # Parse the JavaScript object (simplified approach)
    slideshow_data_str = match.group(1)

    scripts = []

    # Title slide script
    scripts.append({
        'id': 'slide-0',
        'text': "Welcome! This presentation outlines a systematic approach to improving our engineering organization. We'll cover the current challenges and propose a structured framework for implementing AI-powered workflow improvements."
    })

    # Extract script content using regex
    script_matches = re.findall(r'"script":\s*"([^"]+)"', slideshow_data_str)

    for i, script_text in enumerate(script_matches):
        scripts.append({
            'id': f'slide-{i + 1}',
            'text': script_text
        })

    return scripts

def get_api_key():
    """Get API key from the absolute path."""
    key_file_path = '/Users/cam/keys/openai-key.js'
    
    try:
        with open(key_file_path, 'r') as f:
            content = f.read()
        
        # Extract the API key from the JS export
        match = re.search(r'OPENAI_API_KEY\s*=\s*["\'](.*?)["\']', content)
        if match:
            return match.group(1).strip()
        else:
            raise ValueError("Could not find OPENAI_API_KEY in key file")
    except FileNotFoundError:
        raise ValueError(f"Key file not found at {key_file_path}")
    except Exception as e:
        raise ValueError(f"Error reading key file: {e}")

def generate_audio(text, voice='shimmer'):
    """Generate audio using OpenAI TTS API."""
    api_key = get_api_key()

    headers = {
        'Authorization': f'Bearer {api_key}',
        'Content-Type': 'application/json'
    }

    data = {
        'model': 'tts-1',
        'input': text,
        'voice': voice
    }

    data_json = json.dumps(data).encode('utf-8')
    
    request = urllib.request.Request(
        'https://api.openai.com/v1/audio/speech',
        data=data_json,
        headers=headers,
        method='POST'
    )
    
    try:
        with urllib.request.urlopen(request) as response:
            if response.status != 200:
                error_data = response.read().decode('utf-8')
                raise Exception(f"API error {response.status}: {error_data}")
            return response.read()
    except urllib.error.HTTPError as e:
        error_data = e.read().decode('utf-8')
        raise Exception(f"API error {e.code}: {error_data}")

def main():
    """Main function to generate all audio files."""
    print("🎙️  Generating audio files for engineering slideshow...\n")

    # Create audio directory
    audio_dir = Path('slideshow_audio')
    audio_dir.mkdir(exist_ok=True)

    try:
        # Extract scripts
        scripts = extract_slideshow_scripts()
        print(f"📝 Found {len(scripts)} scripts to convert\n")

        # Process each script
        for i, script in enumerate(scripts):
            filename = audio_dir / f"{script['id']}.mp3"

            print(f"[{i + 1}/{len(scripts)}] Generating {script['id']}...")

            if filename.exists():
                print(f"   ✓ Already exists, skipping")
                continue

            try:
                # Generate audio
                audio_data = generate_audio(script['text'])

                # Save to file
                with open(filename, 'wb') as f:
                    f.write(audio_data)

                print(f"   ✓ Generated successfully")

                # Add delay to avoid rate limiting (except for last item)
                if i < len(scripts) - 1:
                    time.sleep(1)

            except Exception as e:
                print(f"   ❌ Error: {e}")
                continue

        # Generate manifest file
        manifest = {
            'voice': 'shimmer',
            'model': 'tts-1',
            'generated': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'files': [
                {
                    'id': s['id'],
                    'filename': f"{s['id']}.mp3",
                    'textLength': len(s['text'])
                }
                for s in scripts
            ]
        }

        with open(audio_dir / 'manifest.json', 'w') as f:
            json.dump(manifest, f, indent=2)

        print(f"\n✅ All audio files generated successfully!")
        print(f"📁 Audio files saved to: {audio_dir}")
        print(f"📋 Manifest file created: {audio_dir}/manifest.json")

    except Exception as e:
        print(f"❌ Error: {e}")
        return 1

    return 0

if __name__ == '__main__':
    # Try to load API key from file
    try:
        get_api_key()
    except ValueError as e:
        print(f"❌ Error: {e}")
        print("\nExpected key file at: /Users/cam/keys/openai-key.js")
        print("The file should contain:")
        print('  export const OPENAI_API_KEY = "your-api-key-here";')
        exit(1)

    exit(main())

File: generate_slideshow_audio.js
#!/usr/bin/env node

const fs = require('fs').promises;
const path = require('path');
const https = require('https');

// Read the slideshow HTML to extract scripts
async function extractSlideshowScripts() {
    const htmlContent = await fs.readFile('engineering_slideshow.html', 'utf-8');
    
    // Extract slideshowData from the HTML
    const match = htmlContent.match(/const slideshowData = ({[\s\S]*?});/);
    if (!match) {
        throw new Error('Could not find slideshowData in HTML');
    }
    
    // Parse the JavaScript object
    const slideshowDataStr = match[1];
    const slideshowData = eval('(' + slideshowDataStr + ')');
    
    // Prepare scripts array
    const scripts = [];
    
    // Title slide script
    scripts.push({
        id: 'slide-0',
        text: "Welcome! This presentation outlines a systematic approach to improving our engineering organization. We'll cover the current challenges and propose a structured framework for implementing AI-powered workflow improvements."
    });
    
    // Content slides scripts
    slideshowData.slides.forEach((slide, index) => {
        scripts.push({
            id: `slide-${index + 1}`,
            text: slide.script
        });
    });
    
    return scripts;
}

// Make API call to OpenAI TTS
async function generateAudio(text, voice = 'shimmer') {
    const apiKey = process.env.OPENAI_API_KEY;
    
    if (!apiKey) {
        throw new Error('OPENAI_API_KEY environment variable is required');
    }
    
    const data = JSON.stringify({
        model: 'tts-1',
        input: text,
        voice: voice
    });
    
    const options = {
        hostname: 'api.openai.com',
        port: 443,
        path: '/v1/audio/speech',
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${apiKey}`,
            'Content-Length': data.length
        }
    };
    
    return new Promise((resolve, reject) => {
        const chunks = [];
        
        const req = https.request(options, (res) => {
            if (res.statusCode !== 200) {
                let errorBody = '';
                res.on('data', chunk => errorBody += chunk);
                res.on('end', () => {
                    reject(new Error(`API error ${res.statusCode}: ${errorBody}`));
                });
                return;
            }
            
            res.on('data', (chunk) => chunks.push(chunk));
            res.on('end', () => resolve(Buffer.concat(chunks)));
        });
        
        req.on('error', (e) => reject(e));
        req.write(data);
        req.end();
    });
}

// Main function
async function main() {
    try {
        console.log('🎙️  Generating audio files for engineering slideshow...\n');
        
        // Create audio directory
        const audioDir = path.join(__dirname, 'slideshow_audio');
        await fs.mkdir(audioDir, { recursive: true });
        
        // Extract scripts
        const scripts = await extractSlideshowScripts();
        console.log(`📝 Found ${scripts.length} scripts to convert\n`);
        
        // Process each script
        for (let i = 0; i < scripts.length; i++) {
            const script = scripts[i];
            const filename = path.join(audioDir, `${script.id}.mp3`);
            
            console.log(`[${i + 1}/${scripts.length}] Generating ${script.id}...`);
            
            try {
                // Check if file already exists
                await fs.access(filename);
                console.log(`   ✓ Already exists, skipping`);
            } catch {
                // File doesn't exist, generate it
                const audioBuffer = await generateAudio(script.text);
                await fs.writeFile(filename, audioBuffer);
                console.log(`   ✓ Generated successfully`);
                
                // Add delay to avoid rate limiting
                if (i < scripts.length - 1) {
                    await new Promise(resolve => setTimeout(resolve, 1000));
                }
            }
        }
        
        console.log('\n✅ All audio files generated successfully!');
        console.log(`📁 Audio files saved to: ${audioDir}`);
        
        // Generate a manifest file for easy reference
        const manifest = {
            voice: 'shimmer',
            model: 'tts-1',
            generated: new Date().toISOString(),
            files: scripts.map(s => ({
                id: s.id,
                filename: `${s.id}.mp3`,
                textLength: s.text.length
            }))
        };
        
        await fs.writeFile(
            path.join(audioDir, 'manifest.json'),
            JSON.stringify(manifest, null, 2)
        );
        
        console.log('📋 Manifest file created: slideshow_audio/manifest.json');
        
    } catch (error) {
        console.error('❌ Error:', error.message);
        process.exit(1);
    }
}

// Check for API key
if (!process.env.OPENAI_API_KEY) {
    console.error('❌ Error: OPENAI_API_KEY environment variable is required');
    console.error('\nTo set it temporarily:');
    console.error('  export OPENAI_API_KEY="your-api-key-here"');
    console.error('\nOr create a .env file in this directory with:');
    console.error('  OPENAI_API_KEY=your-api-key-here');
    console.error('\nThen run: node generate_slideshow_audio.js');
    process.exit(1);
}

// Run the script
main();